### PR TITLE
feat(gateway): introduce `/job-status` endpoint

### PR DESCRIPTION
**Pull request by saurav-malani**
# Description

Introducing two new endpoints, with filters, as a part of improving integration with `rudder-sources`

Endpoints: 
1. `GET /v1/job-status/{job_run_id}`
2. `DELETE /v1/job-status/{job_run_id}`

Filters:
`task_run_id`
`source_id`

## Reference

[Design document](https://www.notion.so/rudderstacks/Design-Proposal-Improved-integration-with-rudder-sources-eb26e569f7404330b44946ea624bb36c#fd8589aa51b2404f8f367e4abed24a04)

## Notion Ticket
https://www.notion.so/rudderstacks/Introduce-job-status-endpoint-s-4c93379144ee46e1b89f82b41454452d

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: gateway/rudder-sources/api/api.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		http.Error(w, "job_id not found", http.StatusBadRequest)
+	}
+
+	err := a.SVC.Delete(ctx, jobId)
+	if err != nil {
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
```
Comment (atzoum) [inline gateway/rudder-sources/api/api.go:61]: Let's make it clear that we are not returning anything
```suggestion
	w.WriteHeader(http.StatusNoContent)
```

File: gateway/rudder-sources/api/api.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	if !ok {
+		http.Error(w, "job_id not found", http.StatusBadRequest)
+	}
+
+	err := a.SVC.Delete(ctx, jobId)
+	if err != nil {
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
```
Comment (atzoum) [inline gateway/rudder-sources/api/api.go:60]: we are not returning anything so no need to set the content-type
```suggestion
```

File: gateway/rudder-sources/model/model.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package model
+
+type JobFilter struct {
+	TaskRunId     *string
```
Comment (atzoum) [inline gateway/rudder-sources/model/model.go:4]: let's turn all filter fields into slices

File: gateway/rudder-sources/api/api_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	"regexp"
+	"testing"
+
+	gomock "github.com/golang/mock/gomock"
+	"github.com/rudderlabs/rudder-server/gateway/rudder-sources/api"
+	"github.com/rudderlabs/rudder-server/gateway/rudder-sources/model"
+	"github.com/stretchr/testify/require"
+)
+
+func TestDelete(t *testing.T) {
```
Comment (atzoum) [inline gateway/rudder-sources/api/api_test.go:17]: can we have tests for the exceptional scenarios as well, e.g. when the service returns an error?

File: gateway/gateway.go [hunk 5]
```diff
@@ -1336,6 +1338,9 @@ func (gateway *HandleT) StartWebHandler(ctx context.Context) error {
 	srvMux.HandleFunc("/v1/failed-events", gateway.fetchFailedEventsHandler).Methods("POST")
 	srvMux.HandleFunc("/v1/clear-failed-events", gateway.clearFailedEventsHandler).Methods("POST")
 
+	sAPI := sources.API{}
+	//rudder-sources new APIs
+	srvMux.PathPrefix("/workspaces").Handler(sAPI.Handler())
```
Comment (atzoum) [inline gateway/gateway.go:1343]: what is the purpose of the `/workspaces` path prefix?

File: gateway/rudder-sources/api/api.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	w.WriteHeader(http.StatusOK)
+
+	body, err := json.Marshal(mapJobToPayload(jobStatus))
+	if err != nil {
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+	_, err = w.Write(body)
+	if err != nil {
```
Comment (atzoum) [inline gateway/rudder-sources/api/api.go:116]: if we failed to write the body to the response writer already, what makes us believe that the `net/http` package will succeed instead?!
Logging an error would be more appropriate here

File: gateway/gateway.go (hunk 5 above)
Comment (saurav-malani) [reply gateway/gateway.go:1343]: that's a mistake. Modified it to `/v1/job-status`.

File: gateway/rudder-sources/api/api.go (hunk 6 above)
Comment (saurav-malani) [reply gateway/rudder-sources/api/api.go:116]: makes sense.

File: gateway/rudder-sources/api/api_test.go (hunk 4 above)
Comment (saurav-malani) [reply gateway/rudder-sources/api/api_test.go:17]: added.

**Review by atzoum — APPROVED** [review]

**Review by mcchran — APPROVED** [review]
LGTM.

File: services/rsources/http/http.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	if !ok {
+		http.Error(w, "job_run_id not found", http.StatusBadRequest)
+	}
+
+	err := h.service.Delete(ctx, jobRunId)
+	if err != nil {
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+	}
+
+	w.WriteHeader(http.StatusNoContent)
```
Comment (mcchran) [inline services/rsources/http/http.go:42]: +1 on taking care of response semantics!

**Pull request merged**